Integraçao

Jonathas-Lopes opened the pull request:
- Configuração de models e schemas para integração ao banco de dados;
- Criação de arquivo de configuração ".env exemplo.json;
- Adição de scheduler para emissão de mensagens websocket e inserção de dados no banco de dados;
- Integração com API do banco central para cotação de moedas;
- Remoção de códigos desnecessários.

Jonathas-Lopes committed:
Criação de template de variavel ambiente

Jonathas-Lopes committed:
Configurando e integrando com Mongodb

Jonathas-Lopes committed:
Desenvolvimento de integração com banco central e agendamento de gravação no banco de dados

Jonathas-Lopes committed:
limpesa de códigos desnecessários

Jonathas-Lopes committed:
mudança de estrutura do projeto

belchior commented on server/.env example.json:
```diff
@@ -0,0 +1,9 @@
+    "DB_PORT": 27017,
+    "Database": "currency"
+  }
+}
```
Recomendo você usar uma lib que leia arquivos `.env`. Dessa forma que você está usando você não conseguiria passar a string de conexão do banco usando uma uma cloud como a AWS ou a Azure.

Dê uma olhada nessa lib: https://www.npmjs.com/package/dotenv

belchior commented on server/entryPoint/server:
```diff
@@ -1,20 +1,23 @@
   },
 });
-
+const connection = require('../src/db/connection/connection');
+connection()
```
Normalmente a conexão com o banco é um processo assíncrono. Dá forma que está sua aplicação pode receber erros de conexão com o banco logo após você fazer o deploy.

Dê uma lida nessa doc para converter para async: https://mongoosejs.com/docs/index.html

belchior commented on server/entryPoint/server:
```diff
@@ -1,20 +1,23 @@
 
 
 http.listen(PORT, function () {
-  console.log("listening on *:3010");
+  console.log(`listening on *:${env.PORT}`);
 });
```
Seria melhor você aguardar a conexão com o banco antes de executar essa linha

belchior commented on server/package.json:
```diff
@@ -3,16 +3,18 @@
   "private": true,
   "scripts": {
-    "start": "nodemon ./bin/www"
+    "start": "nodemon ./entryPoint/server"
```
Existe uma convenção (não é uma regra) de que o script `start` para um servidor seja o comando para executar em mode de produção, para desenvolvimento é usado o termo `dev` ou `start:dev`

belchior commented on server/src/db/connection/connection.js:
```diff
@@ -0,0 +1,9 @@
+const mongoose = require("mongoose")
+const connection = require('../../../.env.json').mongo
+
+
+const connect = function () {
+    mongoose.connect(`${connection.Server}:${connection.DB_PORT}/${connection.Database}`);
+    mongoose.connection.on("error", (error) => console.log(error));
+}
+module.exports = connect;
```
o comentário acima sobre conexão assíncrona vale aqui também

belchior commented on server/src/db/model/currencyModel.js:
```diff
@@ -0,0 +1,67 @@
+        return await CurrencyModel.find({}).limit(50)
+    },
+    insert: async function (currencyNames, currencyObj) {
+        const schemas = currencyNames.map((key, index) => {
+            const values = {
+                'currencyName': currencyObj[key].code,
+                'high': new Array(currencyObj[key].high),
+                'low': new Array(currencyObj[key].low),
+                'date': new Array(currencyObj[key].create_date)
+            }
+            return values
+        })
```
Apenas uma sugestão

```suggestion
        const schemas = currencyNames.map((key) => ({
            'currencyName': currencyObj[key].code,
            'high': new Array(currencyObj[key].high),
            'low': new Array(currencyObj[key].low),
            'date': new Array(currencyObj[key].create_date)
        }))
```

belchior commented on server/src/db/model/currencyModel.js:
```diff
@@ -0,0 +1,67 @@
+            }
+
+        }
+
+
+
+
+
```
Recomendo tu configurar um linter no seu projeto para evitar esses excessos de formatação.

O mais popular em javascript é o eslint: https://eslint.org/docs/user-guide/getting-started

belchior commented on server/src/db/model/currencyModel.js:
```diff
@@ -0,0 +1,67 @@
+            return values
+        })
+
+        for (let schema in schemas) {
```
o termo `schema` é bastante usado para se referir a estrutura de dados relacionada ao banco de dados, recomendo tu usar outro nome.

belchior commented on server/src/db/model/currencyModel.js:
```diff
@@ -0,0 +1,67 @@
+            else {
+
+
+                await CurrencyModel.updateOne(
+                    { currencyName: schemas[schema].currencyName },
+                    { $push: { high: schemas[schema].high } }
+
+
+                );
+                await CurrencyModel.updateOne(
+                    { currencyName: schemas[schema].currencyName },
+                    { $push: { low: schemas[schema].low }, }
+
+
+                );
+                await CurrencyModel.updateOne(
+                    { currencyName: schemas[schema].currencyName },
+                    { $push: { date: schemas[schema].date }, }
+
+
+                );
```
Você poderia usar o `updateMany` nesse caso.

https://mongoosejs.com/docs/api.html#model_Model.updateMany

belchior commented on server/src/integrations/bacen/integration.js:
```diff
@@ -0,0 +1,28 @@
+        
+        const response = await fetch(`https://economia.awesomeapi.com.br/last/${currencies}`);
+        const data = await response.json()
+        currency.insert(Object.keys(data), data)
```
Essa linha pode ser uma possível falha se segurança na sua aplicação! É muito importante você validar todos os dados que venham de uma integração, tem uma lib que ajuda nesse sentido.

https://www.npmjs.com/package/joi

Jonathas-Lopes committed:
Atendendo solicitação de usar biblioteca de variavel de ambiente

Jonathas-Lopes committed:
Atendendo solicitação de tornar conexão com o banco um processo assíncrono

Jonathas-Lopes committed:
Atendendo as seguintes solicitações:
Alteração do método 'map'na função insert;
Alteração do metodo 'updateOne' na função insert;
Validação dos dados da integração com bacen.

Jonathas-Lopes committed:
Atendendo solicitação de adicionar validações devido a falhas de segurança.

Jonathas-Lopes committed:
Atendendo solicitação de aguardar conexão com o banco antes de 'startar' servidor.

Jonathas-Lopes committed:
Adição de libs

Jonathas-Lopes committed:
Atendendo solicitação de adição do sufixo ':dev' no script de inicialização do servidor.

belchior commented on server/src/app.js:
```diff
@@ -1,27 +1,16 @@
 app.use(express.json());
 app.use(express.urlencoded({ extended: false }));
 app.use(cookieParser());
 app.use(express.static(path.join(__dirname, 'public')));
```
Caso você não tenha a intenção de servir arquivos estáticos essa linha é desnecessária

belchior commented on server/.env.example:
```diff
@@ -0,0 +1,4 @@
+SERVER_PORT = 
+DB_HOST = 
+DB_PORT = 
+DB_COLLECTION = 
```
Tu pode usar o arquivo `.env` para guardar as configurações do ambiente de desenvolvimento, em produção essas variáveis serão injetadas via linha de comando que irá sobrescrever os do arquivo

belchior commented on server/entryPoint/server:
```diff
@@ -1,20 +1,24 @@
+  http.listen(PORT, function () {
+    console.log(`listening on *:${PORT}`);
+  });
+})
```
Dessa forma, toda vês que a conexão for reiniciada a sua aplicação vai tentar subir um novo servidor.

Dê uma olhada nesse exemplo de como eu fiz a  conexão com o banco: https://github.com/belchior/learning-graphql/blob/typescript/server/src/server.ts

Jonathas-Lopes replied:
resolvido usando a função 'db.once()'

belchior commented on server/src/db/model/currencyModel.js:
```diff
@@ -0,0 +1,49 @@
+const currency = {
+
+    find: async function () {
+        return await CurrencyModel.find({}).limit(50)
```
Usar o `find` sem passar a um parâmetro não me parece uma boa ideia, e o limite de 50 item está um pouco solto no código.

Tente trocar o nome dessa função por algo mais semântico, algo que já deixe claro a intenção que você quer passar.

Em funções assíncronas não é necessário retornar um await. Ao invês você pode usar `return CurrencyModel.find...`

Jonathas-Lopes replied:
Não vejo outro modo de listar todos os registros de uma vez sem ser com 'find()';
Nome da função trocado;
retirado o 'await' do retorno da função;

belchior commented on server/src/db/model/currencyModel.js:
```diff
@@ -0,0 +1,49 @@
+        }))
+
+        if (!bacenValidator.validate(registers)) {
+            for (let register in registers) {
+
+
+                const currency = await CurrencyModel.findOne({ currencyName: registers[register].currencyName })
+
+
+
+                if (!currency) {
+                    await CurrencyModel.create(registers[register]);
+                }
+                else {
+
+
+                    await CurrencyModel.updateOne(
+                        { currencyName: registers[register].currencyName },
+                        { $push: { high: registers[register].high, low: registers[register].low, date: registers[register].date } }
+
+
+                    )
+                }
```
Fazer queries no banco dentro de um loop é geralmente uma má prática, esse tipo de código gera uma degradação da performance da aplicação.

Tente usar o operador [$in](https://mongoosejs.com/docs/api/query.html#query_Query-in) para trazer todos os `currencies`de uma vês, depois você pode aguardar em um array os item a serem atualizados no banco e usar a função `CurrencyModel.updateMany` para atualizar todos de uma vês.

Jonathas-Lopes replied:
resolvido com 'bulkwrite'

belchior commented on server/src/integrations/bacen/integration.js:
```diff
@@ -0,0 +1,28 @@
+
+        
+        const response = await fetch(`https://economia.awesomeapi.com.br/last/${currencies}`);
+        const data = await response.json()
```
Tu criou o validador não não está usando ele para validar os seus dados

Jonathas-Lopes replied:
Avaliador equivocado

Jonathas-Lopes committed:
Remoção de console log

Jonathas-Lopes committed:
atendendo comentários

Jonathas-Lopes committed:
atendendo comentarios acrescentar dotenv

Jonathas-Lopes committed:
Removendo 'for' das queries, melhorndo a sintaxe das funções, melhorando escopo da função 'listCurrencies.

Jonathas-Lopes committed:
ajuste automático lint

Jonathas-Lopes committed:
correção slint

Jonathas-Lopes committed:
Removendo trechos inuteis de código

Jonathas-Lopes committed:
alteração para emitir uma lista completa apenas na conexão do usuário, depois disto śo haverão emissão de listas diferenciais.

Jonathas-Lopes committed:
alterado para validar todos os dados da API ao invés de um dado por vez.

belchior commented on server/src/db/model/currencyModel.js:
```diff
@@ -0,0 +1,72 @@
+  listCurrencies: async function (aggregate) {
+
+    if (aggregate) {
+      return await CurrencyModel.aggregate([
```
não é necessário usar a keyword `await` quando se pretende retornar uma promise.

 ```suggestion
      return CurrencyModel.aggregate([
```

belchior commented on server/src/db/model/currencyModel.js:
```diff
@@ -0,0 +1,72 @@
+
+    }
+    else
+      return await CurrencyModel.find({});
```
```suggestion
      return CurrencyModel.find({});
```

belchior commented on server/src/integrations/bacen/integration.js:
```diff
@@ -0,0 +1,20 @@
+
+  scheduler.schedule('* * * * *', async function () {
+
+    const response = await fetch(`https://economia.awesomeapi.com.br/last/${currencies}`);
```
É uma boa prática mover o domínio `https://economia.awesomeapi.com.br` para uma variável de ambiente. Em cenário reais você pode ter um domínio para cada ambiente: `test` `stage` e `prod`.

```suggestion
    const { ECONOMY_ENDPOINT } = process.env;
    const response = await fetch(`${ECONOMY_ENDPOINT}/last/${currencies}`);
```

belchior commented on server/src/integrations/bacen/integration.js:
```diff
@@ -0,0 +1,20 @@
+    const data = await response.json();
+    currency.insertCurrency(Object.keys(data), data);
+
+    console.log('listen');
```
tu esqueceu um console.log

belchior commented on server/src/sockets/index.js:
```diff
@@ -1,13 +1,14 @@
 
-    socket.on("disconnect", function () {
-      console.log("A user disconnected");
+    console.log('A user connected');
```
mais um aqui

Jonathas-Lopes committed:
Adicionando eslint

Jonathas-Lopes committed:
Eliminando 'console.log()'

Jonathas-Lopes committed:
movendo domínio de API para arquivo 'dotenv'

Jonathas-Lopes committed:
alteração eslint

Jonathas-Lopes committed:
eliminando 'await' desnecessários nos retornos de funções assincronas

Jonathas-Lopes committed:
Eliminando variaveis desnecessárias.
